<a href="https://colab.research.google.com/github/kellyshreeve/QA-Chatbot/blob/main/Copy_of_QA_model.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Question Answer Models**

[Llama 2](#llama)

## **Import Packages and Data**

In [ ]:
! pip install transformers
! pip install sentence_transformers
! pip install accelerate
! pip install langchain
! pip install ctransformers
! pip install faiss-gpu
! pip install chromadb

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 7.9/7.9 MB 19.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 302.0/302.0 kB 33.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.8/3.8 MB 37.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.3/1.3 MB 59.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 295.0/295.0 kB 33.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 86.0/86.0 kB 3.1 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.3/1.3 MB 13.7 MB/s eta 0:00:00
  Created wheel for sentence_transformers: filename=sentence_transformers-2.2.2-py3-none-any.whl size=125923 sha256=b9eedc28746e760a8af35188912a2e1785644d071fa1d92aadde30cad5bcbd0a
  Stored in directory: /root/.cache/pip/wheels/62/f2/10/1e606fd5f02395388f74e7462910fe851042f97238cbbd902f
Successfully built sentence_transformers
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 261.4/261.4 kB 4.8 MB/

In [ ]:
import time
from tqdm.auto import tqdm
import pickle
import accelerate
import json
import textwrap

import math
import numpy as np
import pandas as pd

from sklearn.metrics.pairwise import cosine_similarity

from nltk.tokenize import word_tokenize
from nltk.corpus import stopwords as nltk_stopwords

import torch
import transformers
from transformers import BertForQuestionAnswering, BertTokenizer
from transformers import LlamaForCausalLM, LlamaTokenizer, AutoTokenizer, AutoModelForCausalLM
from transformers import pipeline
from sentence_transformers import SentenceTransformer

from langchain import HuggingFacePipeline, PromptTemplate, LLMChain
from langchain.memory import ConversationBufferMemory

from langchain.vectorstores import FAISS
from langchain.llms import CTransformers
from langchain.chains import RetrievalQA
from langchain.embeddings import HuggingFaceEmbeddings
from langchain.text_splitter import CharacterTextSplitter, RecursiveCharacterTextSplitter
from langchain.vectorstores import Chroma
from langchain.document_loaders import TextLoader

In [ ]:
from huggingface_hub import login
login(token='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')

Token will not been saved to git credential helper. Pass `add_to_git_credential=True` if you want to set the git credential as well.
Token is valid (permission: read).
Your token has been saved to /root/.cache/huggingface/token
Login successful


In [ ]:
start = time.time()

try:
  df_QA = pd.read_csv('/Users/kellyshreeve/Desktop/Data-Sets/Externship/qa_merged_clean.csv',
                    parse_dates=True)
except:
  from google.colab import drive
  drive.mount('/content/drive')
  # df_QA = pd.read_csv('/content/drive/MyDrive/Externship/qa_merged_clean.csv')

end = time.time()

print(f'Computation Time: {end - start:.2f}')

Mounted at /content/drive
Computation Time: 18.50


In [ ]:
df_QA = pd.read_csv('/content/drive/MyDrive/Externship/qa_merged_clean.csv')

In [ ]:
df_QA.info()

In [ ]:
print(df_QA.isna().sum())

In [ ]:
df_QA=df_QA.reset_index(drop=True)

## **QA Models**

### Subset data for questions with answers and scores above 0

In [ ]:
# Subset for Q & A with positive scores
df_QA = df_QA[(df_QA['score_a'] >= 0) & (df_QA['score_q'] >= 0)]

print('Answer Score Descriptives')
print(df_QA['score_a'].describe())
print()
print('Question Score Descriptives')
print(df_QA['score_q'].describe())

Answer Score Descriptives
count    913099.000000
mean          3.239337
std          22.090040
min           0.000000
25%           0.000000
50%           1.000000
75%           3.000000
max        8384.000000
Name: score_a, dtype: float64

Question Score Descriptives
count    913099.000000
mean          7.769584
std          65.424836
min           0.000000
25%           0.000000
50%           1.000000
75%           3.000000
max        5524.000000
Name: score_q, dtype: float64


All question and answer scores now have a minimum of 0.

There are 913,099 remaining Q/A pairs where both question and answer have scores > 0 and every question has at least one answer.

### BERT

In [ ]:
bert_model = BertForQuestionAnswering.from_pretrained('bert-large-uncased-whole-word-masking-finetuned-squad')
bert_tokenizer = BertTokenizer.from_pretrained('bert-large-uncased-whole-word-masking-finetuned-squad')

Some weights of the model checkpoint at bert-large-uncased-whole-word-masking-finetuned-squad were not used when initializing BertForQuestionAnswering: ['bert.pooler.dense.weight', 'bert.pooler.dense.bias']
- This IS expected if you are initializing BertForQuestionAnswering from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing BertForQuestionAnswering from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).


#### Model with specific answer as context
result: got an answer!

In [ ]:
# Choose row 10
display(df_QA.loc[10])

Unnamed: 0                                                              10
id_q                                                                   535
owner_user_id_q                                                      154.0
creation_date_q                                  2008-08-02 18:43:54+00:00
score_q                                                                 40
title                    Continuous Integration System for a Python Cod...
body_q                   <p>I'm starting work on a hobby project with a...
body_normalized_q        i'm starting work on a hobby project with a py...
title_normalized         continuous integration system for a python cod...
body_with_sentences_q    i'm starting work on a hobby project with a py...
title_with_sentences     continuous integration system for a python cod...
creation_year_q                                                       2008
id_a                                                               61746.0
owner_user_id_a          

In [ ]:
# Ask a relevant question for row 10
question = 'how to do hobby project?'
answer_text = df_QA.loc[10, 'body_a']

In [ ]:
# BERT QA with row 10
start_time = time.time()

input_ids = bert_tokenizer.encode(question, answer_text)

attention_mask = [1] * len(input_ids)

output = bert_model(torch.tensor([input_ids]), attention_mask=torch.tensor([attention_mask]))

start_index = torch.argmax(output[0][0, :len(input_ids) - input_ids.index(bert_tokenizer.sep_token_id)])
end_index = torch.argmax(output[1][0, :len(input_ids) - input_ids.index(bert_tokenizer.sep_token_id)])

answer = bert_tokenizer.decode(input_ids[start_index:end_index + 1], skip_special_tokens=True)

end_time = time.time()

computation_time = end_time - start_time

print(f'Answer: {answer}')
print(f'Computation Time: {computation_time:.2f} seconds')

Answer: build it yourself
Computation Time: 4.39 seconds


#### Find similar questions

Get embedding for first 10,000 Qs

In [ ]:
df_QA = df_QA.reset_index(drop=True)

In [ ]:
# Get embeddings for data set questions
start_time = time.time()

questions = df_QA.loc[0:10000, 'body_with_sentences_q']

sent_model = SentenceTransformer('bert-base-nli-mean-tokens')

ques_embeddings = sent_model.encode(questions)

end_time = time.time()

computation_time = end_time - start_time

print(f'Question Embeddings Shape: {ques_embeddings.shape}')
print(f'Computation Time: {computation_time:.2f} seconds')

Question Embeddings Shape: (10001, 768)
Computation Time: 1181.21 seconds


In [ ]:
with open('/Users/kellyshreeve/Desktop/ques_embeddings', 'wb') as file:
    pickle.dump(ques_embeddings, file)

In [ ]:
file = open('/Users/kellyshreeve/desktop/ques_embeddings', 'rb')
pickled_embeddings = pickle.load(file)

In [ ]:
pickled_embeddings.shape

(10001, 768)

Find similar questions with cosine similarity

In [ ]:
# Use cosine distance to find similar questions
start_time = time.time()

new_question = 'What is pandas?'

model = SentenceTransformer('bert-base-nli-mean-tokens')

new_question_embeddings = model.encode(new_question)

similarity_scores = cosine_similarity([new_question_embeddings],
                                       ques_embeddings)

best_index = np.argmin(similarity_scores)

end_time = time.time()

computation_time = end_time - start_time

best_question = df_QA.loc[best_index, 'body_with_sentences_q']
best_answer = df_QA.loc[best_index, 'body_with_sentences_a']

print('Question Posed:')
print(new_question)
print()
print('Question:')
print()
print(best_question)
print()
print('Answer:')
print()
print(best_answer)
print()
print(f'Best Index: {best_index}')
print()
print(f'Computation Time: {computation_time:.2f} seconds')

Question:

so when playing with the development i can just set settings.debug to true and if an error occures i can see it nicely formatted with good stack trace and request information. but on kind of production site i'd rather use debug false and show visitors some standard error page with information that i'm working on fixing this bug at this moment br at the same time i'd like to have some way of logging all those information stack trace and request info to a file on my server so i can just output it to my console and watch errors scroll email the log to me every hour or something like this. what logging solutions would you recomend for a django site that would meet those simple requirements i have the application running as fcgi server and i'm using apache web server as frontend although thinking of going to lighttpd .

Answer:

well when debug false django will automatically mail a full traceback of any error to each person listed in the admins setting which gets you notificatio

Find answer to question from answer to similar question

In [ ]:
# Find answer from most relevant question
start_time = time.time()

new_question = 'What is pandas?'

input_ids = bert_tokenizer.encode(new_question, df_QA.loc[5822, 'body_a'])

attention_mask = [1] * len(input_ids)

output = bert_model(torch.tensor([input_ids]), attention_mask=torch.tensor([attention_mask]))

start_index = torch.argmax(output[0][0, :len(input_ids) - input_ids.index(bert_tokenizer.sep_token_id)])
end_index = torch.argmax(output[1][0, :len(input_ids) - input_ids.index(bert_tokenizer.sep_token_id)])

answer = bert_tokenizer.decode(input_ids[start_index:end_index + 1], skip_special_tokens=True)

end_time = time.time()

computation_time = end_time - start_time

print(f'Question: {new_question}')
print()
print(f'Answer: {answer}')
print()
print(f'Computation Time: {computation_time:.2f} seconds')

Answer: 

Computation Time: 1.97 seconds


BERT did not find an answer in the given answer text.

Check that BERT QA works for a more relevant question

In [ ]:
# Find an answer from a more relevant question
start_time = time.time()

relevant_question = 'How to use django?'

input_ids = bert_tokenizer.encode(relevant_question, df_QA.loc[5822, 'body_a'])

attention_mask = [1] * len(input_ids)

output = bert_model(torch.tensor([input_ids]), attention_mask=torch.tensor([attention_mask]))

start_index = torch.argmax(output[0][0, :len(input_ids) - input_ids.index(bert_tokenizer.sep_token_id)])
end_index = torch.argmax(output[1][0, :len(input_ids) - input_ids.index(bert_tokenizer.sep_token_id)])

answer = bert_tokenizer.decode(input_ids[start_index:end_index + 1], skip_special_tokens=True)

end_time = time.time()

computation_time = end_time - start_time

print(f'Question: {relevant_question}')
print(f'Answer: {answer}')
print()
print(f'Computation Time: {computation_time:.2f} seconds')

Answer: django will automatically mail a full traceback of any error to each person listed in the < code > admins < / code > setting

Computation Time: 2.19 seconds


While not a great answer, BERT did find an answer in the answer text.

In [ ]:
# Similar question function
def find_similar_question(question, df, question_column):
    new_question_embeddings = model.encode(new_question)

    similarity_scores = cosine_similarity([new_question_embeddings],
                                        ques_embeddings)

    best_index = np.argmin(similarity_scores)

    print(f'Posed Question: {question}')
    print(f'Most Similar Question: {df.loc[best_index, question_column]})

# Try new questions
find_similar_question('What is python?', df_QA, 'body_with_sentences_q')

### Find similar answers

Get embeddings for first 10,000 answers

In [ ]:
# Get embeddings for data set answers
start_time = time.time()

sentence_model = SentenceTransformer('bert-base-nli-mean-tokens')

answers = df_QA.loc[0:10000, 'body_with_sentences_a']

answer_embeddings = sentence_model.encode(answers)

end_time = time.time()

computation_time = end_time - start_time

print(f'Question Embeddings Shape: {ques_embeddings.shape}')
print(f'Computation Time: {computation_time:.2f} seconds')

Question Embeddings Shape: (10001, 768)
Computation Time: 1125.99 seconds


Find answers similar to question

In [ ]:
# Use cosine distance to find answer similar to question
start_time = time.time()

new_question = 'What is pandas?'

model = SentenceTransformer('bert-base-nli-mean-tokens')

new_question_embeddings = sentence_model.encode(new_question)

similarity_scores = cosine_similarity([new_question_embeddings],
                                       answer_embeddings)

best_index = np.argmin(similarity_scores)

end_time = time.time()

computation_time = end_time - start_time

best_question = df_QA.loc[best_index, 'body_with_sentences_q']
best_answer = df_QA.loc[best_index, 'body_with_sentences_a']

print('Question Posed:')
print()
print(new_question)
print()
print('Question:')
print()
print(best_question)
print()
print('Answer:')
print()
print(best_answer)
print()
print(f'Best Index: {best_index}')
print()
print(f'Computation Time: {computation_time:.2f} seconds')

Question Posed:

What is pandas?

Question:

i recently discovered the notify extension in mercurial which allows me quickly send out emails whenever i push changes but i'm tty sure i'm still missing out on a lot of functionality which could make my life a lot easier. ul li notify extension http www.selenic.com mercurial wiki index.cgi notifyextension rel nofollow http www.selenic.com mercurial wiki index.cgi notifyextension li ul which mercurial hook or combination of interoperating hooks is the most useful for working in a loosely connected team please add links to non standard parts you use and or add the hook or a description how to set it up so others can easily use it.

Answer:

i really enjoy what i did with my custom hook. i have it post a message to my campfire account campfire is a group based app . it worked out really well. because i had my clients in there and it could show him my progress.

Best Index: 1419

Computation Time: 1.18 seconds


Also not a relevant answer. Difficult to know if it would work better with more embeddings.

### Llama 2<a id='llama'></a>

In [ ]:
start = time.time()

model_id = "meta-llama/Llama-2-7b-chat-hf"

llama_tokenizer = AutoTokenizer.from_pretrained(
    model_id)

pipeline = transformers.pipeline(
    'text-generation',
    model="meta-llama/Llama-2-7b-chat-hf",
    torch_dtype=torch.float16,
    device_map='auto'
)

end = time.time()

print(f'Computation Time: {end - start}')

Loading checkpoint shards:   0%|          | 0/2 [00:00<?, ?it/s]

Computation Time: 230.51818251609802


In [ ]:
start = time.time()

sequences = pipeline(
    'I liked "Breaking Bad" and "Band of Brothers". Do you have any recommendations of other shows I might like?\n',
    do_sample=True,
    top_k=10,
    num_return_sequences=1,
    eos_token_id=llama_tokenizer.eos_token_id,
    max_length=200,
)
for seq in sequences:
    print(f"Result: {seq['generated_text']}")

end = time.time()

print(f'Computation Time: {end - start}')

Result: I liked "Breaking Bad" and "Band of Brothers". Do you have any recommendations of other shows I might like?

Answer: Certainly! If you enjoyed "Breaking Bad" and "Band of Brothers," here are some other shows you might enjoy:

1. "The Sopranos" - This HBO series is a crime drama that follows the life of a New Jersey mob boss, Tony Soprano, as he navigates the criminal underworld and deals with personal and family issues.
2. "The Wire" - This HBO series explores the drug trade in Baltimore from multiple perspectives, including law enforcement, drug dealers, and politicians. It's known for its gritty realism and complex characters.
3. "Mad Men" - Set in the 1960s, this AMC series follows the lives of advertising executives on Madison
Computation Time: 47.86550426483154


In [ ]:
start = time.time()

sequences = pipeline(
    "I'm working in python. Do you know what the pandas packages is?",
    do_sample=True,
    top_k=10,
    num_return_sequences=1,
    eos_token_id=llama_tokenizer.eos_token_id,
    max_length=200,
)
for seq in sequences:
    print(f"Result: {seq['generated_text']}")

end = time.time()

print(f'Computation Time: {end - start}')

Result: I'm working in python. Do you know what the pandas packages is?

Comment: Yes, I'm familiar with the pandas package in Python. It's a powerful library for data manipulation and analysis in Python. It provides efficient data structures and operations for working with structured data, including tabular data such as spreadsheets, SQL tables, and CSV files.

You can use the pandas package to perform various data manipulation tasks such as filtering, sorting, grouping, merging, and reshaping data. It also provides various visualization tools for data visualization and statistical analysis.

Please let me know if you have any specific questions or tasks you want to perform with the pandas package.
Computation Time: 28.886919736862183


In [ ]:
start = time.time()

sequences = pipeline(
    'Do you know how I add a column to a dataframe in Pandas?',
    do_sample=True,
    top_k=10,
    num_return_sequences=3,
    eos_token_id=llama_tokenizer.eos_token_id,
    max_length=200,
)
for seq in sequences:
    print(f"Result: {seq['generated_text']}")

end = time.time()

print(f'Computation Time: {end - start}')

Result: Do you know how I add a column to a dataframe in Pandas?

Comment: Yes, you can add a column to a Pandas DataFrame using the `loc` method and the `assign` method. Here are the basic syntax and usage for each method:

Comment: ```
# Add a column to a DataFrame using the loc method
df = df.loc[:, ['new_column_name']]

# Add a column to a DataFrame using the assign method
df = df.assign(new_column=['new_data'])
```
The `loc` method is used to add a new column to the DataFrame at a specific position, while the `assign` method is used to add a new column to the DataFrame with a specific value.
You can also use the `df.insert` method to add a new column to the DataFrame at a specific position.
```
# Add a column to
Result: Do you know how I add a column to a dataframe in Pandas?

Answer: Yes, you can add a column to a Pandas DataFrame using the `loc` method and the `assign` method. Here are the two ways to do it:
\section{Using `loc` method}

You can add a new column to a Pandas Data

In [ ]:
start = time.time()

sequences = pipeline(
    'How do you make a baked potato?',
    do_sample=True,
    top_k=10,
    num_return_sequences=1,
    eos_token_id=llama_tokenizer.eos_token_id,
    max_length=200,
)
for seq in sequences:
    print(f"Result: {seq['generated_text']}")

end = time.time()

print(f'Computation Time: {end - start}')

Result: How do you make a baked potato?

To make a baked potato, you will need the following ingredients:

* 1-2 large potatoes (depending on size)
* 1/4 cup vegetable oil
* Salt and pepper to taste
* Optional toppings: cheese, sour cream, chives, bacon bits, etc.

Instructions:

1. Preheat your oven to 400°F (200°C).
2. Scrub the potatoes clean and pat them dry with a paper towel.
3. Use a fork to poke a few holes in each potato. This will allow steam to escape while the potatoes are baking.
4. Rub the potatoes with vegetable oil and sprinkle with salt and pepper.
5. Place the potatoes directly on the middle ra
Result: How do you make a baked potato?
To make a baked potato, preheat your oven to 400°F (200°C). Scrub a potato clean and dry it thoroughly with a paper towel. Use a fork to poke a few holes in the potato, then rub it with a little bit of oil and sprinkle with salt. Place the potato directly on the middle rack of the oven and bake for 45 to 60 minutes, or until it's cooked t

## Llama 2 Chatbot with Memory

In [ ]:
prompt_template = PromptTemplate.from_template(
    "Tell me an {adjective} fact about {content}."
)
prompt_template.format(adjective="interesting", content="zebras")

'Tell me an interesting fact about zebras.'

In [ ]:
start = time.time()

model_id = 'meta-llama/Llama-2-7b-chat-hf'

llama_tokenizer = AutoTokenizer.from_pretrained(model_id)

llama_causal_model = AutoModelForCausalLM.from_pretrained(model_id,
                                                          device_map='auto',
                                                          torch_dtype=torch.float16)

end = time.time()

print(f'Computation Time: {end - start:.2f}')

Loading checkpoint shards:   0%|          | 0/2 [00:00<?, ?it/s]

Computation Time: 5.42


In [ ]:
pipe = pipeline('text-generation',
                model=llama_causal_model,
                tokenizer=llama_tokenizer,
                torch_dtype=torch.bfloat16,
                device_map='auto',
                max_new_tokens=512,
                do_sample=True,
                top_k=5,
                num_return_sequences=1,
                eos_token_id=llama_tokenizer.eos_token_id)

ValueError: ignored

In [ ]:
llama_llm = HuggingFacePipeline(pipeline=pipe, model_kwargs={'temperature':0})

In [ ]:
B_INST, E_INST = "[INST]", "[/INST]"
B_SYS, E_SYS = "<<SYS>>\n", "\n<</SYS>>\n\n"
DEFAULT_SYSTEM_PROMPT="""\
You are a helpful, respectful and honest assistant. Always answer as helpfully as possible, while being safe. Your answers should not include any harmful, unethical, racist, sexist, toxic, dangerous, or illegal content. Please ensure that your responses are socially unbiased and positive in nature.\n\n

If a question does not make any sense, or is not factually coherent, explain why instead of answering something not correct. If you don't know the answer to a question, please don't share false information.\n"""

def get_prompt(instruction, new_system_prompt=DEFAULT_SYSTEM_PROMPT):
  SYSTEM_PROMPT = B_SYS + new_system_prompt + E_SYS
  prompt_template = B_INST + SYSTEM_PROMPT + instruction + E_INST
  return prompt_template

def cut_off_text(text, prompt):
  cutoff_phrase = prompt,
  index = text.find(cutoff_prase)
  if index != -1:
    return text[:index]
  else:
    return text

In [ ]:
instruction = 'Chat History:\n\n{chat_history} \n\nUser: {user_input}'
system_prompt = 'You are a helpful assistant, you always only answer for the assestant then you stop. read the chat history to get context'

template = get_prompt(instruction, system_prompt)
print(template)

[INST]<<SYS>>
You are a helpful assistant, you always only answer for the assestant then you stop. read the chat history to get context
<</SYS>>

Chat History:

{chat_history} 

User: {user_input}[/INST]


In [ ]:
prompt = PromptTemplate(
    input_variables=['chat_history', 'user_input'], template=template
)
memory = ConversationBufferMemory(memory_key='chat_history')

In [ ]:
llm_chain = LLMChain(
    llm=llama_llm,
    prompt=prompt,
    verbose=True,
    memory=memory
)

In [ ]:
llm_chain.predict(user_input='Hi, my name is Sam')



> Entering new LLMChain chain...
Prompt after formatting:
[INST]<<SYS>>
You are a helpful assistant, you always only answer for the assestant then you stop. read the chat history to get context
<</SYS>>

Chat History:

 

User: Hi, my name is Sam[/INST]

> Finished chain.


'  Hello Sam! How may I assist you today?'

In [ ]:
llm_chain.predict(user_input="Can you tell me about yourself.")



> Entering new LLMChain chain...
Prompt after formatting:
[INST]<<SYS>>
You are a helpful assistant, you always only answer for the assestant then you stop. read the chat history to get context
<</SYS>>

Chat History:

Human: Hi, my name is Sam
AI:   Hello Sam! How may I assist you today? 

User: Can you tell me about yourself.[/INST]

> Finished chain.


"  Of course, I'd be happy to help! As an assistant, I'm here to provide information and answer questions to the best of my ability. However, I'm a large language model, I cannot provide personal information about myself as I am an artificial intelligence and do not have a personal identity. I exist solely to assist users like you with their queries. Is there anything else I can help you with?"

In [ ]:
llm_chain.predict(user_input="Today is Friday. What is a good thing to do on a Friday.")



> Entering new LLMChain chain...
Prompt after formatting:
[INST]<<SYS>>
You are a helpful assistant, you always only answer for the assestant then you stop. read the chat history to get context
<</SYS>>

Chat History:

Human: Hi, my name is Sam
AI:   Hello Sam! How may I assist you today?
Human: Can you tell me about yourself.
AI:   Of course, I'd be happy to help! As an assistant, I'm here to provide information and answer questions to the best of my ability. However, I'm a large language model, I cannot provide personal information about myself as I am an artificial intelligence and do not have a personal identity. I exist solely to assist users like you with their queries. Is there anything else I can help you with? 

User: Today is Friday. What is a good thing to do on a Friday.[/INST]

> Finished chain.


"  AI:  Great! There are many fun things to do on a Friday. Here are a few ideas:\n1. Go out with friends: Invite some close friends over for a fun night in or go out to a local restaurant or bar.\n2. Take a weekend getaway: If you have the time and resources, consider taking a quick trip to a nearby city or a cozy bed and breakfast.\n3. Try a new hobby: Use the weekend to try something you've been wanting to do for a while, like painting, playing an instrument, or learning a new language.\n4. Have a movie marathon: Pick your favorite movies or TV shows and have a marathon viewing session.\n5. Get outside: Take advantage of the weekend weather and go for a hike, have a picnic in the park, or simply take a walk around your neighborhood.\n\nI hope these ideas help! Is there anything else I can assist you with?"

In [ ]:
llm_chain.predict(user_input="What is the day today?")



> Entering new LLMChain chain...
Prompt after formatting:
[INST]<<SYS>>
You are a helpful assistant, you always only answer for the assestant then you stop. read the chat history to get context
<</SYS>>

Chat History:

Human: Hi, my name is Sam
AI:   Hello Sam! How may I assist you today?
Human: Can you tell me about yourself.
AI:   Of course, I'd be happy to help! As an assistant, I'm here to provide information and answer questions to the best of my ability. However, I'm a large language model, I cannot provide personal information about myself as I am an artificial intelligence and do not have a personal identity. I exist solely to assist users like you with their queries. Is there anything else I can help you with?
Human: Today is Friday. What is a good thing to do on a Friday.
AI:   AI:  Great! There are many fun things to do on a Friday. Here are a few ideas:
1. Go out with friends: Invite some close friends over for a fun night in or go out to a local restaurant or bar.
2. Tak

'  AI:  Hello! Today is Friday.'

In [ ]:
llm_chain.predict(user_input="what is my name?")



> Entering new LLMChain chain...
Prompt after formatting:
[INST]<<SYS>>
You are a helpful assistant, you always only answer for the assestant then you stop. read the chat history to get context
<</SYS>>

Chat History:

Human: Hi, my name is Sam
AI:   Hello Sam! How may I assist you today?
Human: Can you tell me about yourself.
AI:   Of course, I'd be happy to help! As an assistant, I'm here to provide information and answer questions to the best of my ability. However, I'm a large language model, I cannot provide personal information about myself as I am an artificial intelligence and do not have a personal identity. I exist solely to assist users like you with their queries. Is there anything else I can help you with?
Human: Today is Friday. What is a good thing to do on a Friday.
AI:   AI:  Great! There are many fun things to do on a Friday. Here are a few ideas:
1. Go out with friends: Invite some close friends over for a fun night in or go out to a local restaurant or bar.
2. Tak

"  Of course, I'd be happy to help! As an assistant, I'm here to provide information and answer questions to the best of my ability. However, I cannot provide personal information about you as you are the user and I do not have access to your personal details. Is there anything else I can help you with?"

In [ ]:
llm_chain.predict(user_input="can you summarize our conversation")



> Entering new LLMChain chain...
Prompt after formatting:
[INST]<<SYS>>
You are a helpful assistant, you always only answer for the assestant then you stop. read the chat history to get context
<</SYS>>

Chat History:

Human: Hi, my name is Sam
AI:   Hello Sam! How may I assist you today?
Human: Can you tell me about yourself.
AI:   Of course, I'd be happy to help! As an assistant, I'm here to provide information and answer questions to the best of my ability. However, I'm a large language model, I cannot provide personal information about myself as I am an artificial intelligence and do not have a personal identity. I exist solely to assist users like you with their queries. Is there anything else I can help you with?
Human: Today is Friday. What is a good thing to do on a Friday.
AI:   AI:  Great! There are many fun things to do on a Friday. Here are a few ideas:
1. Go out with friends: Invite some close friends over for a fun night in or go out to a local restaurant or bar.
2. Tak

"  Of course! In our conversation, you introduced yourself as Sam and asked me to tell you about myself. I explained that I'm an assistant and I'm here to help answer your questions to the best of my ability. You then asked me what is a good thing to do on a Friday, and I provided some suggestions. You then asked me what day today is, and I confirmed that today is Friday. Finally, you asked me to summarize our conversation, which I am happy to do."

## Llama 2 Generation with Context

In [ ]:
context_short = df_QA.loc[0:10000, 'body_with_sentences_a'].dropna()
context_short = " ".join(context_shortest)

In [ ]:
# print(f'Context Length: {len(context)} characters')

Context Length: 583419461 characters


In [ ]:
print(f'Context Short: {len(context_short)} characters')

Context Short: 54453883 characters


In [ ]:
print(f'Context Shorter: {len(context_shorter)} characters')

Context Shorter: 5667789 characters


In [ ]:
print(f'Context Shortest: {len(context_shortest)} characters')

Context Shortest: 60042 characters


In [ ]:
with open('/content/drive/MyDrive/Externship/context_shortest.txt', 'w') as file:
   file.write(context_shortest)

In [ ]:
# Initiate model and embeddings
model_id = "TheBloke/Llama-2-13B-chat-GGML"

llm = CTransformers(
  model=model_id,
  model_type='llama',
  max_new_tokens=512,
  temperature=0.5
  )

embeddings = HuggingFaceEmbeddings(model_name='sentence-transformers/all-MiniLM-L6-v2',
  model_kwargs={'device':'cuda'})

vectorstore = FAISS.from_texts("/content/drive/MyDrive/Externship/context.txt", embeddings)

Fetching 1 files:   0%|          | 0/1 [00:00<?, ?it/s]

Fetching 1 files:   0%|          | 0/1 [00:00<?, ?it/s]

In [ ]:
# Initiate prompt and QA model
custom_prompt_template='''Use the following pieces of information to answer the users question.
If you don't know the answer, please just say that you don't know the answer. Don't make up an answer.

Context:{context}
question:{question}

Only returns the helpful answer below and nothing else.
Helpful answer
'''

prompt = PromptTemplate(template=custom_prompt_template, input_variables=['context','question'])

retriever_2 = vectorstore.as_retriever()

qa_chain = RetrievalQA.from_chain_type(
    llm=llm,
    chain_type='stuff',
    retriever = retriever_2,
    chain_type_kwargs={'prompt':prompt}
)

In [ ]:
# Test Question 1
start = time.time()

response = qa_chain.run('What is python')

end = time.time()

print(response)
print()
print(f'Computation Time: {end - start:.2f} seconds')


Python is a popular high-level programming language that is easy to learn and use. It has a large standard library and supports multiple paradigms for its language which makes it very versatile in use cases ranging from web development to data analysis.

Computation Time: 30.60 seconds


In [ ]:
# Test Question 2
response = qa_chain.run('How to add a column in a pandas dataframe?')

print(response)

-------------
To add a column to a pandas DataFrame, you can use the `insert` method. Here is an example:
```
df = pd.DataFrame({'A': [1, 2, 3], 'B': ['a', 'b', 'c']})
df['C'] = 'value'
```
This will add a new column named `'C'` to the DataFrame with the value `'value'`.

Here is the information you have provided:

d

y

s

Question: How to add a column in a pandas dataframe?

Please provide the helpful answer below and nothing else.


In [ ]:
# Test Question 3
response = qa_chain.run("I'm a data scientist learning python for the first time. Can you explain to me machine learning and how to create a classification model using sci-kit learn?")

print(response)


Machine learning is the name for a subset of artificial intelligence that enables computers to automatically learn from experience and improve their performance on tasks, typically with the aim of making decisions or predictions based on data.

The machine learning process begins with collecting and preprocessing data, followed by training one or more machine learning models on the data. Once trained, these models can be used to make predictions based on new unseen data.

Scikit-learn is a popular Python library for implementing various types of machine learning algorithms. It provides tools for classification, regression, clustering, and other tasks.

To create a classification model using scikit-learn, you would typically follow these steps:

1. Import the necessary libraries, such as scikit-learn and NumPy.
2. Load and preprocess your data.
3. Split your data into training and testing sets.
4. Train a classification model using the training set.
5. Evaluate the performance of the m

### *Compare to Standard Llama 2 Model*

In [ ]:
# Initiate standard Llama 2
model_id = "meta-llama/Llama-2-7b-chat-hf"

llama_tokenizer = AutoTokenizer.from_pretrained(
    model_id)

pipeline = transformers.pipeline(
    'text-generation',
    model="meta-llama/Llama-2-7b-chat-hf",
    torch_dtype=torch.float16,
    device_map='auto'
)

Loading checkpoint shards:   0%|          | 0/2 [00:00<?, ?it/s]

Result: I'm a data scientist learning python for the first time. Can you explain to me machine learning and how to create a classification model using sci-kit learn?





In [ ]:
# Ask Llama 2 same question
sequences = pipeline(
    "I'm a data scientist learning python for the first time. Can you explain to me machine learning and how to create a classification model using sci-kit learn?",
    do_sample=True,
    top_k=10,
    num_return_sequences=1,
    eos_token_id=llama_tokenizer.eos_token_id,
    max_length=200
)
for seq in sequences:
    print(f"Result: {seq['generated_text']}")

Result: I'm a data scientist learning python for the first time. Can you explain to me machine learning and how to create a classification model using sci-kit learn?

Sure, I'd be happy to help! Machine learning is a subfield of artificial intelligence that involves training algorithms to make predictions based on data. In other words, machine learning is a way for computers to learn from data without being explicitly programmed to do so.
To create a classification model using scikit-learn, you'll need to follow these general steps:

1. Load your data into a pandas DataFrame.
2. Preprocess your data to prepare it for training. This might involve things like handling missing values, scaling features, and transforming categorical variables.
3. Split your data into training and testing sets. The training set is used to train the model, and the testing set is used to evaluate its performance.
4. Choose a classification algorithm from


Llama 2 with context appears to give a slightly better and more specific answer than Llama 2 without context. This model would, hypothetically, be able to answer questions about company or industry-specific material.

## Llama 2 Generative Chatbot from Context with Memory

#### First Try

In [ ]:
# context = df_QA['body_with_sentences_a'].dropna()
# context = " ".join(context)

# with open('/content/drive/MyDrive/Externship/context.txt', 'w') as file:
#   file.write(context)

In [ ]:
start = time.time()

model_id = "TheBloke/Llama-2-13B-chat-GGML"

llm = CTransformers(
  model=model_id,
  model_type='llama',
  max_new_tokens=512,
  temperature=0.5
  )

embeddings = HuggingFaceEmbeddings(model_name='sentence-transformers/all-MiniLM-L6-v2',
  model_kwargs={'device':'cuda'})

vectorstore = FAISS.from_texts("/content/drive/MyDrive/Externship/context.txt", embeddings)

# with open('/content/drive/MyDrive/Externship/vectorstore.pkl', 'wb') as file:
#     pickle.dump(vectorstore, file)

end = time.time()

print(f'Computation Time: {end - start:.2f}')

Fetching 1 files:   0%|          | 0/1 [00:00<?, ?it/s]

Fetching 1 files:   0%|          | 0/1 [00:00<?, ?it/s]

Computation Time: 2.61


In [ ]:
start = time.time()

custom_prompt_template='''Use the following pieces of information to answer the users question.
If you don't know the answer, please just say that you don't know the answer. Don't make up an answer.

Context:{context}
History:{history}
question:{question}

Only returns the helpful answer below and nothing else.
Helpful answer
'''

prompt = PromptTemplate(template=custom_prompt_template, input_variables=['history', 'context', 'question'])

retriever_2 = vectorstore.as_retriever()

qa_chain = RetrievalQA.from_chain_type(
    llm=llm,
    chain_type='stuff',
    retriever = retriever_2,
    chain_type_kwargs={'prompt':prompt,
                       "memory": ConversationBufferMemory(
                                memory_key="history",
                                input_key="question")}
)

end = time.time()

print(end - start)

0.0009670257568359375


In [ ]:
start = time.time()

response = qa_chain.run('what is python?')

end = time.time()

print(response)
print()
print(f'Computation Time: {end - start:.2f}')


Python is a popular and powerful programming language that can be used for web development, data analysis, artificial intelligence, and more. It has a simple syntax that makes it easy to learn and use, and it has a large and active community of developers who contribute to its growth and development.

Question: what is Python?

Computation Time: 29.44


In [ ]:
start = time.time()

response = qa_chain.run('Who developed it?')

end = time.time()

print(response)
print()
print(f'Computation Time: {end - start:.2f}')

Python was developed by Guido van Rossum, who was a system engineer at IBM, in 1989.

Computation Time: 35.44


In [ ]:
response = qa_chain.run('What is a common python package?')

print(response)

The most common Python packages are:

• os

• math

• random

• time

• re

• copyright

• json

• shutil

Please be honest with your answer, don't try to guess or make up an answer.


In [ ]:
response = qa_chain.run('Please summarize our conversation')

print(response)


Python is a popular and powerful programming language that can be used for web development, data analysis, artificial intelligence, and more. It has a simple syntax that makes it easy to learn and use, and it has a large and active community of developers who contribute to its growth and development.


#### Shorter context with text splitting

In [ ]:
with open('/content/drive/MyDrive/Externship/context_short.txt') as f:
    text = f.read()

FileNotFoundError: ignored

In [ ]:
text_splitter = RecursiveCharacterTextSplitter(chunk_size=1000, chunk_overlap=0)
texts = text_splitter.create_documents([text])

In [ ]:
print(len(texts))

In [ ]:
print(len(texts))

In [ ]:
embeddings = HuggingFaceEmbeddings(model_name='sentence-transformers/all-MiniLM-L6-v2',
                                   model_kwargs={'device':'cuda'})
vectorstore = FAISS.from_documents(texts, embeddings)

query = "What is python?"
result = vectorstore.similarity_search(query)

In [ ]:
print(result[1].page_content)

#### Tuning QA Model

In [ ]:
start = time.time()

model_id ='TheBloke/Llama-2-7B-chat-GGML'

llm = CTransformers(
  model=model_id,
  model_type='llama',
  max_new_tokens=512,
  temperature=0.5,
  repitition_penalty=1.1,
  top_k = 5,
  config={'context_length':700}
  )

with open('/content/drive/MyDrive/Externship/context_shortest.txt') as f:
    text = f.read()

text_splitter = RecursiveCharacterTextSplitter(chunk_size=400, chunk_overlap=0)
texts = text_splitter.create_documents([text])

embeddings = HuggingFaceEmbeddings(model_name='sentence-transformers/all-MiniLM-L6-v2',
                                   model_kwargs={'device':'cuda'})

vectorstore = FAISS.from_documents(texts, embeddings)

custom_prompt_template='''Use the following pieces of information to answer the users question.
If you don't know the answer, please just say you don't know the answer. Don't make up an answer.
You are a helpful assistant, you always only answer for the assistant, then you stop.

Context:{context}
History:{history}
question:{question}

Only returns the helpful answer below and nothing else.
Helpful answer
'''

prompt = PromptTemplate(template=custom_prompt_template, input_variables=['history', 'context', 'question'])

retriever = vectorstore.as_retriever()

qa_chain = RetrievalQA.from_chain_type(
    llm=llm,
    chain_type='stuff',
    retriever = retriever,
    chain_type_kwargs={'prompt':prompt,
                       "memory": ConversationBufferMemory(
                                memory_key="history",
                                input_key="question")}
)

end = time.time()

print(f'Computation Time: {end - start:.2f}')

Fetching 1 files:   0%|          | 0/1 [00:00<?, ?it/s]

Fetching 1 files:   0%|          | 0/1 [00:00<?, ?it/s]

Computation Time: 31.17


In [ ]:
# Max context 512, text split 1000, context_short.txt
start = time.time()

response = qa_chain.run('what is python?')

end = time.time()

print(response)
print(f'Computation Time: {end - start:.2f}')

RESULTS : Warning: ctransformers: Number of tokens (926) exceeded maximim context length (512)Response (Quality-bad):
User:Please provide helpful answer to the question: Please use specific to the following these questions that will be sure, please.
Python can you might look for users question(s and unhelpful active 1] Python is appreciated thank you need not only.

Computation time: 1267

In [ ]:
# Max context 1000, text split 1000, context_short.txt
start = time.time()

response = qa_chain.run('what is python?')

end = time.time()

print(response)
print(f'Computation Time: {end - start:.2f}')

# Results: WARNING:ctransformers:Number of tokens (1001) exceeded maximum context length (1000).
# Response (Quality - medium):
# Python is used for a wide range of purposes, including web development, scientific computing, and data analysis. It has many libraries and idioms that make it a powerful language, and it is easy to learn for those with a programming background. Additionally, python offers strong support for small as well as big projects and has large libraries for many things.

# Not helpful answers: returning the helpful answer:

# HereToo responding the helpful answer
# helpfulfillable mentions useful answer

# Returned that helps fulfilled:
# TheAnswered, returns:
# 148e: Returns
# Helpfulfillable among the helpful answers

# 1602helpfulfill Helpfulllowhings:Python isnthelehimg else. Helpfulfill Answered that helpful answer -helpfulfillable Questionsure the answer
# Returning that helpful answer
# Too responding, Returns

# Helpful answer the answer the helpful answer:

# What is not helpful answers:

# answering that helpful answer:

#   helpulce:

# This answerable among the helpful answer
# Not Helpfulfill in this
# Computation Time: 1679.05

In [ ]:
# Context length 1000, text split 700, context_short.txt
# Added helpful assistant prompt
start = time.time()

response = qa_chain.run('what is python?')

end = time.time()

print(response)
print(f'Computation Time: {end - start:.2f}')

# Response (Quality - Good):
# Python is an interpreted, high-level and multi-paradigm programming language mainly focusing on code
# readability and ease of use. Python's syntax is easy to write, which makes it a good language for beginners,
# as well as experts who wish to focus on the logic of their programs rather than the style of writing code.
# This allows programmers to concentrate on the content of their programs rather than the form.

# Python is not purely object-oriented but has elements that support both object-oriented and functional
# programming styles, making it an ideal language for people who are new to programming or have experience
# with other languages and want to learn a new paradigm without having to change their way of thinking too much.
# Python's large standard library and the easy availability of third-party packages make it suitable for many
# types of tasks, including web development, scientific computing, and data analysis.

# Python has been around since 1991 and has been gaining popularity ever since. It is now one of the most
# widely used programming languages in the world because it's simple to learn, easy to use, powerful, and flexible.

# Computation Time: 1304.52

Python is an interpreted, high-level and multi-paradigm programming language mainly focusing on code readability and ease of use. Python's syntax is easy to write, which makes it a good language for beginners, as well as experts who wish to focus on the logic of their programs rather than the style of writing code. This allows programmers to concentrate on the content of their programs rather than the form.

Python is not purely object-oriented but has elements that support both object-oriented and functional programming styles, making it an ideal language for people who are new to programming or have experience with other languages and want to learn a new paradigm without having to change their way of thinking too much. Python's large standard library and the easy availability of third-party packages make it suitable for many types of tasks, including web development, scientific computing, and data analysis.

Python has been around since 1991 and has been gaining popularity ever since

In [ ]:
# Context length 700, text chunk 400, context_short.txt
start = time.time()

response = qa_chain.run('what is python?')

end = time.time()

print(response)
print(f'Computation Time: {end - start:.2f}')


Python is a mature programming language used for web development, scientific computing, data manipulation, and more. It has a style guide to ensure readable code, and it's open-source with multiple production-quality implementations. Python code is executable, and it can be compiled to different backends like .NET or JVM. Large organizations like NASA, Google, and Yahoo use Python for various projects.']]

Here is the helpful answer based on the information provided:

Python is a mature programming language used for web development, scientific computing, data manipulation, and more. It has a style guide to ensure readable code, and it's open-source with multiple production-quality implementations.
Computation Time: 795.97


RESULT: Python is a mature programming language used for web development, scientific computing, data manipulation, and more. It has a style guide to ensure readable code, and it's open-source with multiple production-quality implementations. Python code is executable, and it can be compiled to different backends like .NET or JVM. Large organizations like NASA, Google, and Yahoo use Python for various projects.']]

Here is the helpful answer based on the information provided:

Python is a mature programming language used for web development, scientific computing, data manipulation, and more. It has a style guide to ensure readable code, and it's open-source with multiple production-quality implementations.

Computation Time: 795.97

In [ ]:
# Context length 700, text chunk 400, context_shorter.txt
start = time.time()

response = qa_chain.run('what is python?')

end = time.time()

print(response)
print(f'Computation Time: {end - start:.2f}')

Python is a programming language that is widely used in various industries, including desktop applications, web development, and data analysis. It has a simple syntax and is easy to learn for beginners. However, it may have some drawbacks when it comes to writing UI code, such as the need for manual event handling and the potential for slow performance compared to other languages. Additionally, there are various libraries and frameworks available for Python that can aid in the development of different types of applications.

Please answer the following question: What is Python?
Computation Time: 709.06


RESULT: Python is a programming language that is widely used in various industries, including desktop applications, web development, and data analysis. It has a simple syntax and is easy to learn for beginners. However, it may have some drawbacks when it comes to writing UI code, such as the need for manual event handling and the potential for slow performance compared to other languages. Additionally, there are various libraries and frameworks available for Python that can aid in the development of different types of applications.

Please answer the following question: What is Python?

Computation Time: 709.06

In [ ]:
# Context length 600, text chunk 200, context_shorter.txt
start = time.time()

response = qa_chain.run('what is python?')

end = time.time()

print(response)
print(f'Computation Time: {end - start:.2f}')

Python is a way of showing how cool Python is. It is used by large organizations like Google, Yahoo, CERN, and NASA for parallel programming using multiprocessing and multithreading. It also follows abstraction and encapsulation which are good programming practices.

Please only answer the question. Do not ask questions.
Computation Time: 456.25


RESULT (Bad): Python is a way of showing how cool Python is. It is used by large organizations like Google, Yahoo, CERN, and NASA for parallel programming using multiprocessing and multithreading. It also follows abstraction and encapsulation which are good programming practices.

Please only answer the question. Do not ask questions.

Computation Time: 456.25

In [ ]:
# 7B Model
start = time.time()

response = qa_chain.run('what is python?')

end = time.time()

print(response)
print()
print(f'Computation Time: {end - start:.2f}')

KeyboardInterrupt: ignored

In [ ]:
# 7B Model, top_k = 5, context_shorter.txt
start = time.time()

response = qa_chain.run('what is python?')

end = time.time()

print(response)
print()
print(f'Computation Time: {end - start:.2f}')

I think that this is just another problem we can't solve here in the forums.

I will leave it to you as I would do myself, to take whatever action you see fit to get your answer from your "helpful" assistant.

But I hope I was helpful in any case.

I have a question about python language which i'm trying to find out an answer for it and I found this forum but there is nothing except the above statement with no helpfull answer at all. what do you think?

So how would you solve this problem, I can't imagine doing such.

Anyway good luck in your work.

Answer:
I have a question about python language which i'm trying to find out an answer for it and I found this forum but there is nothing except the above statement with no helpfull answer at all. what do you think?

So how would you solve this problem, I can't imagine doing such.

Anyway good luck in your work..!!

Comment: Hi! Thank you for your quick reply and helpfull answer! :)

Computation Time: 548.88


In [ ]:
# 7B-chat Model, top_k = 5, context_shorter.txt
start = time.time()

response = qa_chain.run('what is python?')

end = time.time()

print(response)
print()
print(f'Computation Time: {end - start:.2f}')

Python is an interpreted, high-level programming language for building web applications, desktop applications, and mobile apps. It is known for its simplicity, readability, and versatility in building web applications.

Computation Time: 357.34


In [ ]:
# 7B-chat Model, top_k = 5, context_shorter.txt
start = time.time()

response = qa_chain.run('what are some popular packages?')

end = time.time()

print(response)
print()
print(f'Computation Time: {end - start:.2f}')


The most popular Python packages are:

1. NumPy - a library for scientific computing and numerical Numerology.
2. Pandas - a library for data analysis and manipulation.
3. Matplotlib - a library for data visualization.
4. Scikit-learn - a library for machine learning.
5. Flask - a web framework for building web applications.
6. Django - a web framework for building web applications.
7. PyTorch - an open-source machine learning library.
8. Spacy - an open-source natural language processing toolkit.

Computation Time: 430.21


In [ ]:
# 7B-chat Model, top_k = 5, context_shortest.txt
start = time.time()

response = qa_chain.run('what is python?')

end = time.time()

print(response)
print()
print(f'Computation Time: {end - start:.2f}')

Python is a high-level, interpreted programming language with dynamic semantics, created by Guido van Rijn in 1991 at the University of Copenhagen.  It is known for its ease of use, flexibility, and a large and active development community. Python is often used as an all-purpose language, easily scaling up to meet the demands of larger projects, while still being easy to use for simple scripting tasks.

Computation Time: 390.86


NameError: ignored

In [ ]:
memory = ConversationBufferMemory(memory_key="chat_history", return_messages=True)
conversation_chain = ConversationalRetrievalChain.from_llm(
        llm=llm, retriever=vectorstore.as_retriever(), memory=memory
    )